src/system/logger.cc

```diff
@@ -0,0 +1,332 @@
+/*!
+ *
+﻿ * Copyright (C) 2015 Technical University of Liberec.  All rights reserved.
+ *
+ * This program is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License version 3 as published by the
+ * Free Software Foundation. (http://www.gnu.org/licenses/gpl-3.0.en.html)
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+ *
+ *
+ * @file    logger.cc
+ * @brief
+ */
+
+
+#include "system/logger.hh"
+#include "system/global_defs.h"
+#include "config.h"
+
+#include <time.h>
+#include <iomanip>
+
+
+/// Helper function, use for shorten the code point path
+std::string cmn_prefix( std::string a, std::string b ) {
+    if( a.size() > b.size() ) std::swap(a,b) ;
+    return std::string( a.begin(), std::mismatch( a.begin(), a.end(), b.begin() ).first ) ;
+}
+
+
+/*******************************************************************
+ * implementation of LoggerOptions
+ */
+
+LoggerOptions& LoggerOptions::get_instance() {
+	return *instance_;
+}
+
+
+LoggerOptions* LoggerOptions::instance_ = new LoggerOptions();
+
+
+LoggerOptions::LoggerOptions()
+: mpi_rank_(-1), no_log_(false), init_(false) {}
+
+
+LoggerOptions::~LoggerOptions() {
+	if (file_stream_.is_open()) {
+		file_stream_ << std::flush;
+		file_stream_.close();
+	}
+}
+
+
+int LoggerOptions::get_mpi_rank() {
+	return mpi_rank_;
+}
+
+
+int LoggerOptions::setup_mpi(MPI_Comm comm) {
+	ASSERT(!init_).error("Setup MPI must be performed before setting logger file.");
+
+	return MPI_Comm_rank(comm, &mpi_rank_);
+}
+
+
+void LoggerOptions::set_log_file(std::string log_file_base) {
+	ASSERT(!init_).error("Recurrent initialization of logger file stream.");
+
+	if (log_file_base.size() == 0) { // empty string > no_log
+		no_log_ = true;
+	} else {
+		int mpi_rank = LoggerOptions::get_mpi_rank();
+		if (mpi_rank == -1) { // MPI is not set, random value is used
+			std::random_device rd;
+			std::mt19937 gen(rd());
+			std::uniform_int_distribution<int> dis(0, 999999);
+			mpi_rank = dis(gen);
+			WarningOut() << "Unset MPI rank, random value '" << mpi_rank << "' of rank will be used.\n";
+		}
+		std::stringstream file_name;
+		file_name << log_file_base << "." << mpi_rank << ".log";
+		file_stream_.open( file_name.str().c_str(), std::ofstream::out );
+	}
+	init_ = true;
+}
+
+
+void LoggerOptions::reset() {
+	mpi_rank_ = -1;
+	no_log_ = false;
+	init_ = false;
+	if (file_stream_.is_open()) {
+		file_stream_ << std::flush;
+		file_stream_.close();
+	}
+}
+
+
+/*******************************************************************
+ * implementation of StreamMask
+ */
+
+
+StreamMask StreamMask::cout_mask()
```

Comment:
1. Try to make static variables for these (since it is more a constant then a getter)
2. remove _mask postfix
3. rename 'file_mask' to 'log'